packages/next/pages/_document.js

```diff
@@ -134,12 +129,26 @@ export class Head extends Component {
     return <head {...this.props}>
       {children}
       {head}
-      {page !== '/_error' && <link rel='preload' href={`${assetPrefix}/_next/static/${buildId}/pages${pagePathname}${_devOnlyInvalidateCacheQueryString}`} as='script' nonce={this.props.nonce} crossOrigin={this.props.crossOrigin || process.crossOrigin} />}
+      {amphtml && <>
+        <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1"/>
+        <link rel="canonical" href={asPath === '/amp' ? '/' : asPath.replace(/\/amp$/, '')} />
+        {/* https://www.ampproject.org/docs/fundamentals/optimize_amp#optimize-the-amp-runtime-loading */}
+        <link rel="preload" as="script" href="https://cdn.ampproject.org/v0.js" />
+        {/* Add custom styles before AMP styles to prevent accidental overrides */}
+        {styles && <style amp-custom="" dangerouslySetInnerHTML={{__html: styles.map((style) => style.props.dangerouslySetInnerHTML.__html)}} />}
+        <style amp-boilerplate="" dangerouslySetInnerHTML={{__html: `body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}`}}></style>
+        <noscript><style amp-boilerplate="" dangerouslySetInnerHTML={{__html: `body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}`}}></style></noscript>
+        <script async src="https://cdn.ampproject.org/v0.js"></script>
+      </>}
+      {!amphtml && <>
```

Comment:
Does this check whether the route actually exist as opposed to assuming that there is such a `/amp` path?

Reply:
The initial implementation is "dumb" in that it links between amp / non amp assuming that both exist / are valid. This made me think we might have to introduce a way to mark a page being able to render to amp, instead of having `.amp.js`. Btw not using `.amp.js` will be faster with compilation too. Downside being we might have to introduce something like:

```jsx
function Index() {

}

Index.getInitialProps = () => {
	return {
		foo: 'bar'
	}
}

Index.ampReady = true // Not sold on the naming, but this is illustrational.

export default Index
```